engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSinkFactoryV2.scala

```diff
@@ -0,0 +1,152 @@
+package pl.touk.nussknacker.engine.avro.sink
+
+import cats.data.Validated
+import cats.data.Validated.{Invalid, Valid}
+import pl.touk.nussknacker.engine.api.context.ProcessCompilationError.{CustomNodeError, NodeId}
+import pl.touk.nussknacker.engine.api.context.transformation.{DefinedEagerParameter, NodeDependencyValue}
+import pl.touk.nussknacker.engine.api.context.{ProcessCompilationError, ValidationContext}
+import pl.touk.nussknacker.engine.api.definition._
+import pl.touk.nussknacker.engine.api.process.{ProcessObjectDependencies, SinkFactory}
+import pl.touk.nussknacker.engine.api.typed.typing.{TypedObjectTypingResult, TypingResult}
+import pl.touk.nussknacker.engine.api.{LazyParameter, MetaData}
+import pl.touk.nussknacker.engine.avro.schemaregistry.SchemaRegistryProvider
+import pl.touk.nussknacker.engine.avro.{KafkaAvroBaseTransformer, SchemaDeterminerErrorHandler}
+import pl.touk.nussknacker.engine.avro.KafkaAvroBaseTransformer.{SchemaVersionParamName, SinkKeyParamName, SinkValidationModeParameterName, SinkValueParamName, TopicParamName}
+import pl.touk.nussknacker.engine.avro.typed.AvroSchemaTypeDefinitionExtractor
+import pl.touk.nussknacker.engine.flink.api.process.FlinkSink
+import pl.touk.nussknacker.engine.api.typed
+import pl.touk.nussknacker.engine.definition.parameter.editor.ParameterTypeEditorDeterminer
+
+object KafkaAvroSinkFactoryV2 {
+
+  private val restrictedParamNames = Set(SchemaVersionParamName, SinkKeyParamName, SinkValidationModeParameterName, TopicParamName)
+
+  private def containsRestrictedNames(obj: TypedObjectTypingResult): Boolean =
+    (obj.fields.keySet & restrictedParamNames).nonEmpty
+
+  private def toParamEither(typing: TypingResult)(implicit nodeId: NodeId): Validated[ProcessCompilationError, Either[List[Parameter], Parameter]] = {
+    def toSingleParam(name: String, typing: TypingResult) =
+      Parameter(name, typing).copy(
+        isLazyParameter = true,
+        // TODO
+        editor = new ParameterTypeEditorDeterminer(typing).determine()
+      )
+
+    def toObjectFieldParams(typedObject: TypedObjectTypingResult) =
+      typedObject.fields.map { case (name, typing) =>
+        toSingleParam(name, typing)
+      }.toList
+
+    def toParamEither(typing: TypingResult, isRoot: Boolean)(implicit nodeId: NodeId) =
+      typing match {
+        case typed.typing.Unknown =>
+          Invalid(CustomNodeError(nodeId.id, "Cannot determine typing for provided schema", None))
+        case typedObject: TypedObjectTypingResult if containsRestrictedNames(typedObject) =>
+          Invalid(CustomNodeError(nodeId.id, s"""Record field name is restricted. Restricted names are ${restrictedParamNames.mkString(", ")}""", None))
+        case typedObject: TypedObjectTypingResult if isRoot =>
+          Valid(Left(toObjectFieldParams(typedObject)))
+        case _: TypedObjectTypingResult =>
```

Comment:
Can we add handling "simple" nested elements? Or at least add comment for that?

Reply:
I've added handling of nested fields